### DISCUSSION TITLE
Pressing super in full screen mode produces a dummy panel

### DISCUSSION THREAD

**Discussion by ComposerShirley** [linuxmint/discussions#359] [Ideas]
The intelligent auto-hide panel feature in Cinnamon is great, definitely superior to Win. The only issue I see is that if you press 'super' to access the panel/menu (I do this often in full screen mode while teaching) you can select items from the menu, but not from the panel. I'm guessing this is just a little oversight.

**EDIT:** I updated the title after discovering this issue was universal and unrelated to the intelligent auto-hide panel feature. Original title: _Pressing super with "intelligent auto-hide panel" should produce a useable panel_

**Comment by LincolnFarrell** [comment]
Came here to make a post discussing this, but I guess you've already done that. Always felt like a bug to me not being able switch tabs like that, so I hope this feature is implemented.

**Comment by cplittlfield** [comment]
The menu and the panel are separate applets.  Activating the menu with the Super key does **not** display the panel as well, unless you have that setting enabled in the system settings (a feature that I understand was added to satisfy Windows users who just couldn't cope).

In general, however, the point of hitting Super is to open and use the menu, and not to use the panel.  If I want to see the panel, it's because I want to click on something, so why not open the panel with the mouse?  (I could support the addition of a shortcut to open the panel application with a keystroke combination, but honestly, what would be the point?  We'd still need to use the mouse in any case.)

The logic of the Windows interface is considerably muddled at this point.  After a year and a half of using Linux Mint, I find its logic refreshingly clear and easy to grasp.  If I wanted to continue operating exactly the way Windows does, I'd have stayed with Windows.

**Comment by ComposerShirley** [comment]
Windows does not have an intelligent auto-hide panel (toolbar) feature like Cinnamon. This is one of many areas where Linux Mint is superior. I understand that the menu and panel are two separate applets, as you say, however when using the intelligent auto-hide panel feature, pressing the super key does indeed bring up both the menu _and_ the panel (which is logical). Since it brings up both, it would only be logical to be able to interact with both.

When running programs in full screen mode the panel isn't accessible with the mouse (with intelligent auto-hide panel enabled). And that is as it should be. So ideally, accessing the panel with the super key would allow one to not only _see_ it, but also _use_ it, without having to exit full-screen mode.

If you dislike Cinnamon's unique intelligent auto-hide panel feature, that's a different topic.

**Reply by cplittlfield** [reply]
That's very interesting, because on my system, even with a maximized app on the monitor where the panel is, I can cause the panel to be displayed by bringing the cursor to the edge of the screen.  This is true with either "auto hide" or "intelligently hide" selected.  So I'm not sure what is going on with your system.

Also, when you press the Super key, the panel is displayed only if the Menu applet is configured to display the panel (Menu > Panel > Force the panel to be visible when opening the menu > On).  Otherwise, it does not display the panel.

I would, as previously mentioned, support adding the ability to display the panel using a keyboard shortcut, even though I don't need it on my system.

**Reply by ComposerShirley** [reply]
Not when an app is maximized, but when it's in full screen mode (often F11). For example in Celluloid, Firefox, etc.

> Also, when you press the Super key, the panel is displayed only if the Menu applet is configured to display the panel (Menu > Panel > Force the panel to be visible when opening the menu > On). Otherwise, it does not display the panel.

Sure, but I believe they are tethered by default. But either way it would be nice if the panel worked when displayed. Being able to assign a shortcut would be an improvement, but in this case it would be a work around to fixing what feels like a simple oversight.

**Reply by cplittlfield** [reply]
The are not "tethered," precisely.  But it does seem as though what you are asking for is for the two separate applications to be integrated into one, not just a way to open the panel.  Good luck.

**Comment by ComposerShirley** [comment]
Actually, I just tested it, the behavior is the same whether using intelligent auto-hide panel or the default, always show the panel. So, even in the default panel mode, if you are running a program in full screen mode, pressing 'super' brings up both the menu _and_ the panel, but while you can interact with the menu, you can't interact with the panel.

**Reply by cplittlfield** [reply]
As mentioned, you _should_ be able to get the panel to show itself by bringing the cursor to the edge of the screen where the panel is located.  If that's not working for you, I can see why you would want some way to display the panel and work with it.  But displaying the panel along with the menu is not the designed way to interact with the panel at present.

I'm simply suggesting that it might be more appealing to the devs to ask for a keyboard shortcut to activate the panel when it is hidden (there may even be one already, but I haven't found it).  Such a shortcut might be easier to implement than trying to integrate the menu and panel applets.

**Comment by spamegg1** [comment]
I've seen many forum posts about this and I directed users over here.
Let me repeat my view on the Forums here in this discussion as well.

I've been using Cinnamon since *before* the "Force the Panel to be visible when opening Menu" option was added:
![Screenshot from 2024-07-22 00-39-07](https://github.com/user-attachments/assets/b29f189b-5cf1-407f-99d7-b73d44e7ca86)

Now, this was done because many users requested it.
And these users were mostly newcomers from Windows.
They had ingrained habits where they pressed the "Windows key" to make their Panel visible.
They also had habits to use the Windows key as an "escape hatch from fullscreen".
This is especially true for gamers, who are used to "freely alt-tabbing in an out of fullscreen games" on Windows.
This was always a strange thing on Windows, that the Menu and the Panel were somehow forcefully "attached" to each other.
But long ingrained habits defeat logic and reason easily...

However I think Mint developers misunderstood the motivation behind the requests.
So, currently with this option enabled, opening the Menu forces the Panel to be visible, but that's it.
That's not the reason users asked for it.
Adding to this, the fact that many games, when run on Linux, do not have the same "alt-tabbing behavior" (sometimes they have window focus issues due to Wine + how they are coded to run on Windows), and users were frustrated.

So I think that Cinnamon should either:

go all the way and fully embrace the Windows behavior (where Panel + Menu are "merged" and act as one, and it acts as a "fullscreen escape hatch"), 

or revert back to the old behavior and remove the currently available (and arguably useless) option.

**Reply by spamegg1** [reply]
I vote for the second option.

Cinnamon's original view was that the Menu and Panel are separate. The menu is just an applet that can be added to / removed from the Panel, no different than the others. It should work independently from the Panel.

**Reply by billyswong** [reply]
In MATE desktop, if I fullscreen / F11 the Firefox, then press the Super key to trigger the MintMenu, the panel is fully accessible. So this is not a Windows-only feature. I see it not about the menu being "merged" with the panel, but the menu being a child element of the panel. So it makes sense for the parent to become both visible and accessible when one visits the child, sort of like the panel become the foreground "window" instead of the fullscreen program, in order to let the menu surface appear in foreground.

**Reply by spamegg1** [reply]
Hmm I didn't know that about MATE, thanks.
MATE being a continuation of Gnome 2, which was very 90's Windows-inspired, makes sense I guess.

I don't agree with the parent / child analogy, but that's more of a philosophical debate I think.
I think the main problem is "how much Windows habits do we want to implement in Cinnamon to cater to newcomers' ingrained habits".
I voted for the second option to steer away from that direction.
Users will demand more and more "exactly like Windows" things.
Again, I'm fine with the first option. Anything is better than the current "in-between" state.

**Reply by cplittlfield** [reply]
Billyswong's parent/child analogy makes sense to me, but I also agree with spamegg about creeping Windows-ism.  It is clear that the Mint team sees the menu and the panel as separate apps, and I see no particular reason for them to have to change, even if the MATE devs see things differently.  On the other hand, merging the menu and the panel might be an approach that would make sense to Clem and the others, I don't know.

It seems to me that another factor to consider is what the precise difference is between full-screen mode and maximizing the window.  Is there a reason for thinking of full-screen as something that takes over the screen and makes accessing the panel, etc., difficult?  Should it be like that, or should it be simply a borderless, menu-less version of a maximized window?  That question seems to be the real issue for the OP.  I'm not a gamer, so I never even knew that full-screen and maximized behave differently.  Might that be the real issue to address?  (And if so, how?)

**Reply by ComposerShirley** [reply]
I'm not a gamer either, but as a teacher and someone who gives other presentations/performances, full-screen mode is very important. Many programs, especially ones that are commonly used for presentations, have a full-screen mode. If you're showing a film for example, you don't want anything displayed other than the film itself. If you're teaching, you don't want students distracted by seeing tools, scroll bars, "File", "Edit", strewn about. Some kind of clean maximized window likely wouldn't make sense, because it would mean rewriting all the programs that already have a full-screen mode...reinventing the wheel.

If there were another keyboard shortcut, or a way to assign a shortcut to display a usable panel, that would work, but I think it's more logical for the super key to bring up both the menu and panel. Is there really a case where you would want the super key to bring up the menu alone? It seems to me that if you call up the menu, there's no harm in showing a usable panel too.

Would that specific behavior be like Windows? Sure, but there are plenty of things about Cinnamon that are like Windows. It seems as though the concept is take the good, improve when possible, leave the bad, and forge new paths when it makes sense. Does it make sense to avoid Windows-isms for the sake of avoiding Windows-isms?

If another better, non-Windows-ish solution is thought of, then fantastic. I'll be the first to want to learn it. But in the meantime it might make sense to just fix this one little oversight, since the developers already took the path they did.

**Reply by billyswong** [reply]
@cplittlfield
Fullscreen mode hides not only the border of the window itself. It also instructs the window manager to put the window in front of and cover the panel(s) so that the panel(s) won't distract audience who are not supposed to know or care offtopic information such as which OS version and theme the presenter is using in the computer.

Fullscreen mode is more important for those who do public/group presentation or film playing, than those who play games personally. One can get used to the panel(s) of one's own computer and not be distracted by them.

Also, due to the long existence of super key as fullscreen mode escape without telling the application to exit fullscreen mode, software applications generally don't provide a built-in function for presenters to escape fullscreen mode yet not expose all the other content behind the scene of their presentations, as that would be reinvent the wheel.

**Reply by cplittlfield** [reply]
I don't use full-screen mode for much of anything, so thank you both for the discussion of the need for it.

I keep my panel auto-hidden and have never needed to use it, on those rare occasions when I have put an application into full-screen mode, so I was not aware that the panel is inaccessible in that mode.  Another reason I never noticed that the panel might not be accessible is that I have two monitors, and if an app is in full-screen mode on monitor 2, the panel is still available on monitor 1.

But I also have to confess that I am having trouble imagining circumstances in which it might be necessary to see either the panel or the menu while showing a video to students.  Isn't that disruptive, not to mention very distracting to the students?  On the other hand, I can easily see myself fiddling with the keyboard or mouse during a video and displaying the menu or panel completely by accident, so in my case, the current behavior of full-screen mode is probably a good thing, lol! :smile:

**Reply by ComposerShirley** [reply]
I don't play films for students, but I'll give you an example of something I do while teaching that's been a bit frustrating:

I teach music, so I run MuseScore quite a bit in full-screen mode so that, 1) I can use every pixel of screen real estate space to make the music as large as possible ao the kids can read it, and, 2) It's critical that I remove everything on the screen except the music as a class of kids would be completely distracted by anything extra at all.

But I'm not just using MuseScore. I often need to periodically display a PDF or two (again maximizing space/minimizing distractions), and often I need to switch over to a program to playback MP3s or MIDI recordings. Having to exit/reenter full-screen in order to switch programs adds quite a few steps, can mess up formatting, and adds distractions/unprofessionalism.

**Reply by cplittlfield** [reply]
Ah, that makes a lot of sense.  I'm so used to switching with Alt-Tab that I generally don't even think of using the panel for that.  Sorry for the lack of imagination on my part. :smiley_cat:

**Reply by ComposerShirley** [reply]
Yes, alt-tab is definitely a good option in many cases (speaking of Windows-isms!). Of course sometimes I'll want to load a program that isn't already running, with a click on a panel icon. But thanks for the reminder, making alt-tab more of a habit will help.

**Comment by kisielo85** [comment]
the fact that this issue is still not resolved is bizarre.
If the panel is visible, and my mouse is visible, clicking on it should just work.

Luckily after some tinkering with Cinnamon's code I managed to get it working by editing  2 lines in `/usr/share/cinnamon/js/ui/popupMenu.js`

here is my repo with a simple script applying the patch: [github.com/kisielo85/Menu-focus-tweak-for-Cinnamon](https://github.com/kisielo85/Menu-focus-tweak-for-Cinnamon)

**Reply by ComposerShirley** [reply]
Thanks so much. This works perfectly.

**Reply by tytyereal** [reply]
This is perfect! I looked so long for this solution, thank you. This should definitely be implemented into base cinnamon, honestly it should be how it works by default because as you said "If the panel is visible, and my mouse is visible, clicking on it should just work". But even as a setting to toggle on this would be greatly appreciated

**Reply by catcherofafly** [reply]
Wow thank you :) coming from windows i can't understand why this isn't standard in mint. Finally the panel is clickable in fullscreen like it should be :D

**Reply by jazix** [reply]
Hey @kisielo85,
nice fix!
~~But i noticed it also blocks keyboard events and therefore prevents typing in the opened menu after pressing the super key.
I almost only open programs that way, so i really missed that function with your fix.
I made a modification, so the code only overrides the behavior of non keyboard events.
This way your fix stays in place while the menu can still be used as usual.~~

edit: turns out it was just my system messing around.
@kisielo85 's fix works perfectly fine!

**Reply by MefixRU** [reply]
I also have a solution for some games. I focused on the "borderless window" mode, since the problem only occurs in some. I created a separate extension with settings. It mimics Windows behavior as closely as possible. I spent a day struggling with it and was about to give up on switching to Linux. I hope this helps someone else too. https://github.com/MefixRU/hide-panel-cinnamon-for-games/tree/main

**Comment by ArqamQazi** [comment]
Fixed with https://github.com/linuxmint/cinnamon/commit/c709a6c1384d42ceacf2e0ff68e0e5ed1f704280

**Reply by MefixRU** [reply]
Is this fixed in 22.3 Cinnamon 6.6.9?

I installed it a week ago 22.3 Cinnamon 6.6.9 and ran into a problem with Dota 2 game focus not deactivating in borderless window mode. 
I tried various settings, and smart hide didn't help; the panel would automatically reappear when I moved the mouse close to it. I even created an extension that makes the behavior in problematic games as close as possible to Windows behavior when pressing the win(super) button.

**Comment by mtwebster** [comment]
This is fixed in Cinnamon and will be in Cinnamon 6.8. There are no plans to backport it (there are far too many other changes surrounding this to do so.
